lib/queue.js

```diff
@@ -798,22 +755,38 @@ class Queue extends Emitter {
    *   exception executing the batch or readying the connection.
    * @modifies {arguments}
    */
-  saveAll(jobs) {
-    return this._commandable().then((client) => {
-      const batch = client.batch(),
-        errors = new Map();
-
-      for (const job of jobs) {
-        try {
-          job
-            ._save((evalArgs) => this._evalScriptOn(batch, evalArgs))
-            .catch((err) => void errors.set(job, err));
-        } catch (err) {
-          errors.set(job, err);
-        }
+  async saveAll(jobs) {
+    const client = await this._commandable();
+    const batch = client.pipeline(),
+      errors = new Map(),
+      defers = [];
+
+    const final = Promise.all(
+      jobs.map((job) => {
+        const defer = helpers.defer();
+        return job
+          ._save((evalArgs) => {
+            this._evalScriptOn(batch, evalArgs);
+            defers.push(defer);
+            return defer.promise;
+          })
+          .catch((err) => {
+            // Catch both errors from defer.reject below (from the EVALSHA), and
+            // from any serialization failures.
+            errors.set(job, err);
+          });
+      })
+    );
+    const results = await batch.exec();
+    for (const [defer, [error, jobId]] of helpers.zip(defers, results)) {
+      if (error) {
+        defer.reject(error);
+      } else {
+        defer.resolve(jobId);
       }
-      return helpers.callAsync((done) => batch.exec(done)).then(() => errors);
-    });
+    }
+    await final;
```

Comment:
`ioredis` has abysmal support for pipelining as the pipeline commands don't return promises but instead return the pipeline for chaining. This means we have to manually identify which of the jobs each command corresponds to :\